### PR TITLE
Switch to multi-asset coin selection algorithm

### LINKED ISSUE #1416: `Examples.ECDSA` (and probably `Schnorr`) test is fragile on a non-private cluster
The test is not specific enough when trying to spend the output of a validator, and may select a UTxO with no datum, which CTL does not allow for scripts.

The script lies at the following address on preview, and has a UTxO [(`0@3ca8852ff76537383dded28f3d65249b76b605890193ee306f7e0584b2e5d19a`) with no datum](https://preview.cexplorer.io/address/addr_test1wqkzgx3nvmayxl4m06lyym83gv3twclnypqm8jnm62za4hgnygzfj/utxo#data).

Note that the `Schnorr` test has the same behaviour, although does not have a UTxO with no datum, so does not yet encounter this issue.

### PR DESCRIPTION
**Pull request by errfrom**
Closes https://github.com/Plutonomicon/cardano-transaction-lib/issues/1131
Closes https://github.com/Plutonomicon/cardano-transaction-lib/issues/1349
Closes #1416 

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [ ] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: src/Internal/BalanceTx/CoinSelection.purs
Comment (t4ccer) [inline src/Internal/BalanceTx/CoinSelection.purs:2]: Dead link, use links pinned to a commit

**Review by jy14898 — CHANGES_REQUESTED** [review]
I'll take a look at the rest tomorrow

**Review by jy14898 — CHANGES_REQUESTED** [review]
I'm pretty happy with this, I've cross checked the implementation to the ones linked in the comments as much as I can and it seems to be correct

File: src/Internal/BalanceTx/BalanceTx.purs
Comment (jy14898) [reply src/Internal/BalanceTx/BalanceTx.purs:265]: It's mostly about how it reads:

```purescript
((\tx -> mkBalancerState tx ...) >>> prebalanceTx)
```

wouldn't imply to me that the tx becomes fully balanced. I understand `prebalanceTx` calls the next transition in the machine after it's prebalanced, but it's not clear from that naming and I have to go and inspect the implementation.

IMO runNextBalancingStep could do the recursive calls, decisions on next steps. and checking termination conditions. After which, we're effectively back to having a main loop function.

It's mostly about naming, and it's difficult to name the function, which maybe means it does too much?

Comment (errfrom) [reply src/Internal/BalanceTx/BalanceTx.purs:267]: Ok now I see what you mean, will try to refactor it

**Review by jy14898 — APPROVED** [review]
LGTM, all my previous concerns have been addressed

In terms of testing (not sure if that will be a different PR so approving now), I think at least some golden tests for utxo distribution after a series of transactions might be nice. That'd at least protect us from accidental subtle changes to coin selection by requiring explicit approval.

**Review by jy14898 — APPROVED** [review]

**Review by Bradley-Heather — APPROVED** [review]

**Comment by Bradley-Heather** [comment]
Please merge in develop again and fix any breaking tests now that CI is up and running.

**Comment by errfrom** [comment]
Please do not merge - some Plutip tests fail due to some bug in the balancer.

**Review by jy14898 — CHANGES_REQUESTED** [review]
Blocking merge due to bug

**Comment by errfrom** [comment]
**Update**:
1. **[Plutip tests]** Fixed a bug where the change would not be generated if the pre-computed change `Value`s contained no tokens and the amount of Ada required to cover a minimum Ada value exceeded the available Ada amount. 
(commit https://github.com/Plutonomicon/cardano-transaction-lib/pull/1268/commits/44570053cedc5cbe16f937e2fa938bd21021636f)
2. **[Staking tests]** Fixed a bug where the balancer could create transactions without inputs, causing an error when submitting such a transaction. (commit https://github.com/Plutonomicon/cardano-transaction-lib/pull/1268/commits/0bcd4395b9e884819ba6fbbd94b7e5453e00f06f)
3. E2E tests are still failing (see https://github.com/Plutonomicon/cardano-transaction-lib/issues/1335)

File: src/Internal/BalanceTx/Error.purs
Comment (jy14898) [inline src/Internal/BalanceTx/Error.purs:15]: This error is no longer thrown, but mentioned in the FAQ. I'm guessing it's replaced by `BalanceInsufficientError`?

**Review by jy14898 — CHANGES_REQUESTED** [review]

**Comment by errfrom** [comment]
@Bradley-Heather @klntsky @jy14898 Do we have plans to include the new coin selection as part of the next release? I think we should merge this PR into `develop` to be able to gather some user feedback.

File: src/Internal/BalanceTx/BalanceTx.purs
Comment (klntsky) [inline src/Internal/BalanceTx/BalanceTx.purs:368]: Why transaction and not balancedTx (the updated one)?

File: src/Internal/Types/ScriptLookups.purs
Comment (klntsky) [inline src/Internal/Types/ScriptLookups.purs:1090]: Why don't we need to throw here anymore?

Comment (jy14898) [reply src/Internal/Types/ScriptLookups.purs:1090]: Wallet addresses can contain datum

**Pull request merged**
